components/script/dom/bindings/str.rs

```diff
@@ -541,6 +566,69 @@ fn parse_date_component(value: &str) -> Result<(u32, u32, u32), ()> {
     Ok((year_int, month_int, day_int))
 }
 
+/// https://html.spec.whatwg.org/multipage/#parse-a-time-component
+fn parse_time_component(value: &str) -> Result<(u32, u32, f32), ()> {
+    // Step 1
+    let mut iterator = value.split(':');
+    let hour = iterator.next().ok_or(())?;
+    if hour.len() != 2 {
+        return Err(());
+    }
+    let hour_int = hour.parse::<u32>().map_err(|_| ())?;
+
+    // Step 2
+    if hour_int > 23 {
+        return Err(());
+    }
+
+    // Step 3, 4
+    let minute = iterator.next().ok_or(())?;
+    if minute.len() != 2 {
+        return Err(());
+    }
+    let minute_int = minute.parse::<u32>().map_err(|_| ())?;
+
+    // Step 5
+    if minute_int > 59 {
+        return Err(());
+    }
+
+    // Step 6
+    let mut second_float: f32 = 0.0;
+    if let Some(second) = iterator.next() {
+        second_float = second.parse::<f32>().unwrap();
```

Comment:
Why unwrap here?

Reply:
it need to unwrap to get value.
second_float = second.parse::<f32>()
> the trait `bitflags::<unnamed>::convert::From<bitflags::<unnamed>::num::ParseFloatError>` is not implement

Reply:
My real question is why do we not use ? here?

Reply:
because `second_float = second.parse::<f32>()?;` causes error

Reply:
How about `.map_err(|_| ())?` ?

Reply:
I try. not work.
How about `let num = "12.34#5".parse::<f32>().unwrap_or_else(()); `

Reply:
How about something like this?

```rust
fn parse_time_component(value: &str) -> Result<(u32, u32, f32), ()> {
    // Step 1
    let mut iterator = value.split(':');
    
    let hour_int = match iterator.next() {
        Some(hour) => {
            if hour.len() != 2 {
                return Err(());
            }

            hour.parse::<u32>().map_err(|_| ())?
        },
        None => return Err(())
    };

    // Step 2
    if hour_int > 23 {
        return Err(());
    }

    // Step 3, 4
    let minute_int = match iterator.next() {
        Some(minute) => {
            if minute.len() != 2 {
                return Err(());
            }

            minute.parse::<u32>().map_err(|_| ())?
        },
        None => return Err(())
    };

    // Step 5
    if minute_int > 59 {
        return Err(());
    }

    // Step 6
    let second_float = match iterator.next() {
        Some(second) => second.parse::<f32>().map_err(|_| ())?,
        None => return Err(())
    };

    Ok((hour_int, minute_int, second_float))
}
```

Reply:
This is fine. 
```
 let second_float = match iterator.next() {
        Some(second) => second.parse::<f32>().map_err(|_| ())?,
        None => return Err(())
    };
```

but why not `.parse::<f32>().unwrap_or(());`?

Reply:
OK. This is good in the end!

```
if let Some(second) = iterator.next() {
  second_float = second.parse::<f32>().map_err(|_| ())?
}
```